# Main entrypoint of the workflow. 
# Please follow the best practices: 
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there. 
import os
import re
import glob
from snakemake.utils import min_version

min_version("7.18.2")

configfile: "config/config.yaml"

# list all subfolders in input_dir
sample_dirs = os.listdir(config['input_dir'])

include: "rules/01-concatenate_fastq.smk"
include: "rules/02-filtering.smk"
# add:
# - strip primers/barcodes/adapters
# - need alignment first to align 3' ends before clustering?
# orient reads to the same strand
# dereplicate
# remove singletons (must with nanopore!)
include: "rules/03-clustering.smk"
include: "rules/04-mapping.smk"
include: "rules/05-polish_racon.smk"
include: "rules/06-relabel.smk"
include: "rules/07-clustering_identity.smk"
include: "rules/08-taxonomy.smk"
include: "rules/09-fix_otu_table.smk"
include: "rules/10-prep_for_ampvis2.smk" # <- shouldn't be necessary, amp_load() reads sintax files directly, see "taxonomy" here https://kasperskytte.github.io/ampvis2/reference/amp_load.html#arguments-1
include: "rules/11-prep_for_phyloseq.smk"
rule all:
    input:
        #expand("results/final/{id}/OTUtable_tax_{id}_blast.tsv", id=["97", "99"]), #<- final output, let's see if it works - works for un-merged compressed files
        #expand("results/final/{id}/phyloseq_tax_{id}_blast.tsv", id=["97", "99"]), #<- final output, let's see if it works - works for un-merged compressed files
        #expand("results/final/{id}/phyloseq_abundance_{id}_blast.tsv", id=["97", "99"])
        expand(os.path.join(config['tmp_dir'], "samples", "{sample}_filtered.fastq"), sample=sample_dirs),
        expand(os.path.join(config['output_dir'], "final", "report", "total_reads.tsv"))
