# Main entrypoint of the workflow.
# Please change the rule all to to your needs (blast or sintax or both, 97 or 99% similarity, etc.)
# You can get both sintax and blast annotation by adding the corresponding output in the rule all.
import os
import re
from snakemake.utils import min_version

min_version("7.18.2")

configfile: "config/config.yaml"

# list all subfolders in input_dir
sample_dirs = os.listdir(config['input_dir'])

include: "rules/01-concatenate_fastq.smk"
include: "rules/02-filtering.smk"
include: "rules/03-clustering.smk"
include: "rules/04-mapping.smk"
include: "rules/05-polish_racon.smk"
include: "rules/06-relabel.smk"
include: "rules/07-clustering_identity.smk"
include: "rules/08-taxonomy.smk"
include: "rules/09-fix_otu_table.smk"
include: "rules/10-prep_for_ampvis2.smk" # <- shouldn't be necessary, amp_load() reads sintax files directly, see "taxonomy" here https://kasperskytte.github.io/ampvis2/reference/amp_load.html#arguments-1
include: "rules/11-prep_for_phyloseq.smk"

## Outputs for blast
#expand("results/final/{id}/OTUtable_tax_{id}_blast.tsv", id=["97", "99"]), #<- Final output, tax + abundance for ampvis2 using blast annotation
#expand("results/final/{id}/phyloseq_tax_{id}_blast.tsv", id=["97", "99"]), #<- Taxonomy output for phyloseq using blast annotation
#expand("results/final/{id}/phyloseq_abundance_{id}_blast.tsv", id=["97", "99"]), #<- Abundance output for phyloseq using blast annotation
rule all:
    input:
        expand(os.path.join(config['output_dir'], "final", "{id}", "OTUtable_tax_{id}_sintax.tsv"), id=["97", "99"]),
        expand(os.path.join(config['output_dir'], "final", "{id}", "phyloseq_tax_{id}_sintax.tsv"), id=["97", "99"]),
        expand(os.path.join(config['output_dir'], "final", "{id}", "phyloseq_abundance_{id}_sintax.tsv"), id=["97", "99"]),
        expand(os.path.join(config['output_dir'], "final", "report", "total_reads.tsv"))